---
title: "Cardiac in vitro models transcriptomics (Onodi et al., 2022)"
author: "Juan C. Sanchez-Arias"
date: "16/03/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

[Onodi et al., 2022](https://doi.org/10.1016/j.yjmcc.2021.12.007) described the similarities and differences in gene expression profiles between human and murine cardiac myocyte cell lines and human adult cardiac tissue using bulk RNA-seq. The data has been deposited in ArrayExpress and it is available for mining. Overall, cardiac myocyte cell lines do not fully recapitulate primary cardiomyocyte tissue (as expected); however, differentiated hiPSCs and neonatal cardiomyocyte (at 3 *days-in-vitro*) are decent models for their respective systems.

## 1. Set up

Load necessary packages for subsequent analysis

```{r other-packages, include=TRUE, warning=FALSE}

#General Bioconductor packages
    library(Biobase)
    library(oligoClasses)
     
#Annotation and data import packages
    library(ArrayExpress)
    library(pd.hugene.1.0.st.v1)
    library(hugene10sttranscriptcluster.db)
     
#Quality control and pre-processing packages
    library(oligo)
    library(arrayQualityMetrics)
     
#Analysis and statistics packages
    library(limma)
    library(topGO)
    library(DESeq2)
    library(ReactomePA)
    library(clusterProfiler)
     
#Plotting and color options packages
    library(gplots)
    library(ggplot2)
    library(geneplotter)
    library(RColorBrewer)
    library(pheatmap)
    library(enrichplot)
     
#Formatting/documentation packages
   #library(rmarkdown)
   #library(BiocStyle)
    library(dplyr)
    library(tidyr)

#Helpers:
    library(stringr)
    library(matrixStats)
    library(genefilter)
    library(openxlsx)

```

## 2. Downloading data from ArrayExpress

Onodi et al. (2022) deposited data per species analyzed as shown below:

-   *Homo sapiens*: [E-MTAB-10720](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-10720/)
-   *Rattus norvegicus*: [E-MTAB-10913](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-10913/)
-   *Mus musculus*: [E-MTAB-10912](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-10912/)

```{r raw-data-check, echo=TRUE}

data_dir <- tempdir()

if (!dir.exists(data_dir)) {
    dir.create(data_dir)
}
```

The function `getAE` downloads the data into the temporary directory (`data_dir`).

```{r raw-data-import, include=TRUE, message=FALSE, warning=FALSE}

anno_AE <- getAE("E-MTAB-10913", path = data_dir)
```

Each dataset in ArrayExpress is stored following MAGE-TAB (MicroArray Gene Expression Tabular) specifications:

-   Investigation Description Format (IDF): contains top level information about the experiment including title, description, submitter contact details and protocols
-   Array Design Format (ADF)
-   Sample and Data Relationship Format (SDRF): contains essential information on the experimental samples, e.g. the experimental group(s) they belong to.
-   Raw data files
-   Processed data files

## 3. Creating ExpressionSet with `Biobase`

An `ExpressionSet` can be manipulated (e.g., subsetted, copied), and is the input to or output of many Bioconductor functions.

The data in an `ExpressionSet` consist of:

-   assayData: Expression data from microarray experiments with microarray probes in rows and sample identifiers in - columns
-   metaData
-   phenoData: A description of the samples in the experiment with sample identifiers in rows and description elements in columns; holds the content of the SDRF file
-   featureData: metadata about the features on the chip or technology used for the experiment with same rows as assayData by default and freely assignable columns
-   further annotations for the features, for example gene annotations from biomedical databases (annotation). experimentData: A flexible structure to describe the experiment.

> The functions `pData` and `fData` extract the sample and feature annotation, respectively, from an ExpressionSet. The function `exprs` will return the expression data itself as a matrix.

> The data deposited was normalized with DESeq2 (for more details regarding pre-processing see protocol [P-MTAB-113851](https://www.ebi.ac.uk/arrayexpress/protocols/1420244/?ref=E-MTAB-10913))

```{r SDRF, include=TRUE}

sdrf_location <- file.path(data_dir, anno_AE$sdrf)
SDRF <- read.delim(sdrf_location)

rownames(SDRF) <- SDRF$Assay.Name
# SDRF <- AnnotatedDataFrame(SDRF)
```


```{r matrix-count-to-DESeq2}
processedData_location <- file.path(data_dir, anno_AE$processedFiles)

cts <- read.delim(processedData_location,comment.char="#", row.names = "Geneid")

cts <- cts[6:20]
```

Creating an object with sample information table `coldata` (e.g., treatment, type, etc.)

```{r sample-info-object-01, include=TRUE}

celltype <- c("H9C2-undiff", "H9C2-undiff", "H9C2-undiff",
              "H9C2-RA", "H9C2-RA", "H9C2-RA",
              "NRCDM-DIV1", "NRCDM-DIV1", "NRCDM-DIV1",
              "NRCDM-DIV3", "NRCDM-DIV3", "NRCDM-DIV3",
              "adult rat heart", "adult rat heart", "adult rat heart"
)

dev.stage <- c("embryo", "embryo", "embryo",
               "embryo", "embryo", "embryo",
               "neonate", "neonate", "neonate",
               "neonate", "neonate", "neonate",
               "adult", "adult", "adult"
    
)
    
coldata <- data.frame(celltype, dev.stage, row.names = colnames(cts))

coldata$celltype <- factor(coldata$celltype, levels = c("H9C2-undiff",
                                                        "H9C2-RA",
                                                        "NRCDM-DIV1", 
                                                        "NRCDM-DIV3", 
                                                        "adult rat heart"))
```

> It is absolutely critical that the columns of the count matrix and the rows of the column data (information about samples) are in the same order. DESeq2 will not make guesses as to which column of the count matrix belongs to which row of the column data, these must be provided to DESeq2 already in consistent order.

```{r sample-info-object-01, include=TRUE, echo=TRUE}
all(rownames(coldata) %in% colnames(cts))
```
With the count matrix, `cts`, and the sample information, `coldata`, we can construct a `DESeqDataSet`:

```{r DESeqDataSet-01, include=TRUE}
dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ celltype)

dds_h9c2 <- DESeqDataSetFromMatrix(countData = cts[1:6],
                              colData = subset(coldata, 
                                               celltype == "H9C2-undiff" | celltype == "H9C2-RA"),
                              design = ~ celltype)

dds_nrcdm <- DESeqDataSetFromMatrix(countData = cts[7:12],
                              colData = subset(coldata, 
                                               celltype == "NRCDM-DIV1" | celltype == "NRCDM-DIV3"),
                              design = ~ celltype)
```

## 4. Pre-filtering

While it is not necessary to pre-filter low count genes before running the DESeq2 functions, there are two reasons which make pre-filtering useful: by removing rows in which there are very few reads, we reduce the memory size of the dds data object, and we increase the speed of the transformation and testing functions within DESeq2. ere we perform a minimal pre-filtering to keep only rows that have at least 10 reads total. Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function.

```{r pre-filtering, include=TRUE}
keep <- rowSdDiffs(counts(dds)) >= 10
dds <- dds[keep,]

keep_h9c2 <- rowSdDiffs(counts(dds_h9c2)) >= 10
dds_h9c2 <- dds_h9c2[keep_h9c2,]

keep_nrcdm <- rowSdDiffs(counts(dds_nrcdm)) >= 10
dds_nrcdm <- dds_nrcdm[keep_nrcdm,]
```

## 5. Note on factor levels

By default, R will choose a reference level for factors based on alphabetical order. Then, if you never tell the DESeq2 functions which level you want to compare against (e.g. which level represents the control group), the comparisons will be based on the alphabetical order of the levels. There are two solutions: you can either explicitly tell results which comparison to make using the contrast argument (this will be shown later), or you can explicitly set the factors levels. In order to see the change of reference levels reflected in the results names, you need to either run `DESeq` or `nbinomWaldTest`/`nbinomLRT` after the re-leveling operation. Setting the factor levels can be done in two ways, either using `factor` or using `relevel`:

```{r factor-levels, include=TRUE}
# dds$celltype <- factor(dds$celltype, 
#                        levels = c(
#                            #"adult rat heart", 
#                            "H9C2-undiff", 
#                            "H9C2-RA" 
#                            #"NRCDM-DIV1",
#                            #"NRCDM-DIV3"
#                            ))

# dds$celltype <- factor(ddsdes$celltype,
#                        levels = c(
#                            "adult rat heart",
#                            "H9C2-undiff",
#                            "H9C2-RA",
#                            "NRCDM-DIV1",
#                            "NRCDM-DIV3"
#                            ))
```

## 6. Differential expression analysis

The standard differential expression analysis steps are wrapped into a single function, `DESeq.` The estimation steps performed by this function are described below, in the manual page for `?DESeq` and in the Methods section of the `DESeq2` publication (Love, Huber, and Anders 2014).

```{r de-analysis, include=TRUE}


dds <- DESeq(dds)

dds_h9c2 <- DESeq(dds_h9c2)
# res <- results(dds)
# res

dds_nrcdm <- DESeq(dds_nrcdm)
res_nrcdm <- results(dds, contrast= c("celltype", "NRCDM-DIV3","NRCDM-DIV1"))

res <- results(dds, contrast= c("celltype", "H9C2-RA","H9C2-undiff"))
res
```
## 7. p-values and adjusted p-values

We can order our results table by the smallest *p* value:

```{r p-value-sort, include=TRUE}
resOrdered <- res[order(res$pvalue),]
resOrdered_nrcdm <- res_nrcdm[order(res_nrcdm$pvalue),]
```

We can summarize some basic tallies using the `summary` function.

```{r summary-res, include=TRUE}
summary(res)
```
### Log fold change shrinkage for visualization and ranking

```{r lfc, include=TRUE}
library(apeglm)

resLFC <- lfcShrink(dds, coef="celltype_H9C2.RA_vs_H9C2.undiff", type="apeglm")
resLFC
```


## 8. Exploring and exporting results

### MA-plot

In `DESeq2`, the function `plotMA` shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the `DESeqDataSet.` Points will be colored red if the adjusted p value is less than 0.1. Points which fall out of the window are plotted as open triangles pointing either up or down.

```{r ma-plot-01, include=TRUE}
plotMA(res, ylim=c(-2,2))
```
```{r ma-plot-01, include=TRUE}
plotMA(resLFC, ylim=c(-2,2))
```
## 9. Save table

```{r write, include=TRUE}
write.csv(as.data.frame(resOrdered), 
          file="celltype_groups_H9c2Undiff-RA.csv")

write.csv(as.data.frame(resOrdered_nrcdm), 
          file="celltype_groups_nrcdm.csv")
```

## 10. Data transformation and visualization

### Count data transformations

DESEq2 operates on raw counts and use discrete distributions to test for differential expression. 

For other downstream analyses – e.g. for visualization or clustering – it might be useful to work with transformed versions of the count data.

Maybe the most obvious choice of transformation is the logarithm. Since count values for a gene can be zero in some conditions (and non-zero in others), some advocate the use of pseudocounts

Alternatives transformations:

- *Variance stabilizing transformations (VST)*
- *Regularized logarithm or rlog*

### Blind dispersion estimation

### Extracting transformed values

```{r transformation-01, include=TRUE}
vsd <- vst(dds, blind=FALSE)
rld <- rlog(dds, blind=FALSE)
head(assay(vsd), 3)
```
```{r}

vsd_h9c2 <- vst(dds_h9c2, blind=FALSE)
rld_h9c2 <- rlog(dds_h9c2, blind=FALSE)
head(assay(vsd_h9c2), 3)

```

### Variance stabilizing transformation

### Regularized log transformation

### Effects of transformations on the variance

```{r transformation-02, include=TRUE}
# this gives log2(n + 1)
ntd <- normTransform(dds)
library("vsn")
meanSdPlot(assay(ntd))
```

## 11. QC

### Heatmap of the count matrix

```{r qc-01}
library("pheatmap")
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds)[,c("celltype")])
rownames(df) <- colnames(ntd)
pheatmap(assay(ntd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)
```
```{r}
pheatmap(assay(rld)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)
```
```{r}
plotPCA(vsd, intgroup=c("celltype"))
```
```{r}
write.csv(as.data.frame(assay(vsd)), 
          file="celltype_groups_vsd.csv")
```

```{r}
write.csv(as.data.frame(assay(dds)), 
          file="celltype_groups_dds.csv")
```

```{r}
write.csv(as.data.frame(assay(rld)), 
          file="celltype_groups_rld.csv")
```

```{r}
library(EnhancedVolcano)

 EnhancedVolcano(res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue')
```

