---
title: "GSE99551 analysis"
author: "Juan C. Sanchez-Arias"
date: "14/12/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Dataset information
GEO Accession ID: [GSE99551](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE99551).
- Title: "Gene expression analysis of WT and Adam19KO cardiac neural crest cells".
- Contributors: Arai H, Yamamoto T, Sehara-Fujisawa A.
- Associated paper: Arai HN, Sato F, Yamamoto T, Woltjen K et al. Metalloprotease-Dependent Attenuation of BMP Signaling Restricts Cardiac Neural Crest Cell Fate. Cell Rep 2019 Oct 15;29(3):603-616.e5. [PMID: 31618630](https://www.ncbi.nlm.nih.gov/pubmed/31618630).
- Files available:
  - GPL6246-18741.txt
  - GPL6246.annot
  - GPL6246.txt
  - log2_rma_values.txt

```{r libraries, include=TRUE, warning=FALSE, message=FALSE, echo=FALSE}
library(tidyverse)
```

## Reading and tidying data

```{r loading-data-01, include=TRUE, warning=FALSE, message=FALSE}
library(readr)
GPL6246_anno <- read_delim("GPL6246.txt", 
    delim = "\t", escape_double = FALSE, 
    col_types = cols_only(ID = col_guess(),
                          `Gene symbol` = col_guess()), trim_ws = TRUE)

GPL6246_values <- read_delim("log2_rma_values.txt", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE)

## Annotating data frame with gene symbols
GPL6246_anno_in <- GPL6246_anno %>% mutate(ma_present = GPL6246_anno$ID %in% GPL6246_values$ID_REF)

GPL6246_anno_in <- GPL6246_anno_in %>% filter(!GPL6246_anno_in$ma_present == FALSE)

GPL6246_values_anno <- GPL6246_values %>% 
  mutate(gene_symbol = case_when(ID_REF == GPL6246_anno_in$ID ~ GPL6246_anno_in$`Gene symbol`))
```

## Getting LQTS genes

```{r results-01, include=TRUE}
GPL6246_lqts <- GPL6246_values_anno %>% filter(toupper(gene_symbol) == "ANK2"|
                                    toupper(gene_symbol) == "KCNQ1"|
                                    toupper(gene_symbol) == "KCNH2"|
                                    toupper(gene_symbol) == "SCN5A"|
                                    toupper(gene_symbol) == "KCNE1"|
                                    toupper(gene_symbol) == "KCNE2"|
                                    toupper(gene_symbol) == "KCNJ2"|
                                    toupper(gene_symbol) == "CACNA1C"|
                                    toupper(gene_symbol) == "CAV3"|
                                    toupper(gene_symbol) == "SCN4B"|
                                    toupper(gene_symbol) == "AKAP9"|
                                    toupper(gene_symbol) == "SNTA1"|
                                    toupper(gene_symbol) == "ALG10")
```

## Time to plot!

```{r plot-01, include=TRUE}

GPL6246_lqts %>% ggplot(aes(reorder(gene_symbol, VALUE, sum), VALUE, fill = reorder(gene_symbol, VALUE, sum))) + 
  geom_col() +
  scale_y_continuous(limits = c(0,11), breaks = seq(0, 11, by = 1)) +
  scale_fill_manual(values = rev(troy)) +
  geom_hline(aes(yintercept = median(GPL6246_lqts$VALUE)), linetype = "dashed") +
  theme_bw() +
  theme(legend.position = "none") +
  ylab(label = "log2(expression value)") +
  xlab(label = "LQTS gene") +
  labs(title = "Transcript log2 expression values of\nLQTS associated genes in E14.5 cardiac neural crest cells", 
       caption = "Source: Deposited data by Arai et al., 2019") +
  theme(axis.text.y = element_text(face = "italic"),
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 12)) +
  geom_label(aes(2, median(df_wt_long$log2_expression, digits = 2),
                 label = paste0("Median value: ", 
                                round(median(GPL6246_lqts$VALUE), digits = 2)),
                 hjust = -0.15),
             fill = "white") +
  coord_flip()
```

